pkg/guestagent/guestagent_linux.go

```diff
@@ -244,3 +246,30 @@ func (a *agent) Info(ctx context.Context) (*api.Info, error) {
 	}
 	return &info, nil
 }
+
+const deltaLimit = 2 * time.Second
+
+func (a *agent) fixSystemTimeSkew() {
+	ticker := time.NewTicker(10 * time.Second)
+	defer ticker.Stop()
+	for now := range ticker.C {
+		rtc, err := timesync.GetRTCTime()
+		if err != nil {
+			logrus.Warnf("fixSystemTimeSkew: lookup error: %s", err.Error())
+			continue
+		}
+		d := rtc.Sub(now)
+		logrus.Debugf("fixSystemTimeSkew: rtc=%s systime=%s delta=%s",
+			rtc.Format(time.RFC3339), now.Format(time.RFC3339), d)
+		if d > deltaLimit || d < -deltaLimit {
+			err = timesync.SetSystemTime(rtc)
+			if err != nil {
+				logrus.Warnf("fixSystemTimeSkew: set system clock error: %s", err.Error())
+				continue
+			}
+			logrus.Infof("fixSystemTimeSkew: system time synchronized with rtc")
+			break
+		}
+	}
+	go a.fixSystemTimeSkew()
```

Comment:
What's this recursion?

Reply:
I assumed this is due to a concern that the time adjustment could cause the ticker not to trigger again. E.g. when you move back the clock by one hour, will the next tick happen in 10s or 1h and 10s?

So any time the clock is adjusted, the current goroutine exits and a new instance sets up a new ticker based on the current clock settings.

I don't know if this is necessary, and I don't think the clock will be adjusted backwards, but it seems like sensible defensive coding to me.

Reply:
Doesn't this cause infinite recursion?

Reply:
> Doesn't this cause infinite recursion?

Depends what you mean by that. It is an endless loop, but there are only at most 2 instances at any point in time: the routine loops forever, unless there is a time delta. If there is no, it adjusts the time, starts a new instance of itself, and then stops, which in turn will also remove the current ticker.

In the end, there is just a single check every 10s to see if there is time drift. If there is time drift every single time, then the time will be adjusted every 10s and the job passed on to the next instance. But conceptually this isn't different from an endless loop.

I guess given the confusion this is causing, using a nested loop for this would be better (but functionally equivalent):

```
func timeskew {
    forever {
        every 10 seconds {
            if time drift {
                adjust time
                break
            }
        }
    }
}
```

The current implementation is using "recursion" instead of the outer loop:

```
func timeskew {
    every 10 seconds {
        if time drift {
            adjust time
            break
        }
    }
    call timeskew // with tailcall recursion optimization
}
```

Reply:
Since the current implementation uses goroutines, it is not really recursion because the old instance terminates just as the new instance starts. Which also triggers the `defer` statements to `Stop` the old ticker, so there is no resource leakage.

Reply:
I've not tested it, but this would be the version without "recursion":

```diff
$ g diff -b
diff --git pkg/guestagent/guestagent_linux.go pkg/guestagent/guestagent_linux.go
index b4721b5..22e1e31 100644
--- pkg/guestagent/guestagent_linux.go
+++ pkg/guestagent/guestagent_linux.go
@@ -250,8 +250,8 @@ func (a *agent) Info(ctx context.Context) (*api.Info, error) {
 const deltaLimit = 2 * time.Second

 func (a *agent) fixSystemTimeSkew() {
+       for {
                ticker := time.NewTicker(10 * time.Second)
-       defer ticker.Stop()
                for now := range ticker.C {
                        rtc, err := timesync.GetRTCTime()
                        if err != nil {
@@ -271,5 +271,6 @@ func (a *agent) fixSystemTimeSkew() {
                                break
                        }
                }
-       go a.fixSystemTimeSkew()
+               ticker.Stop()
+       }
 }
```

Reply:
@AkihiroSuda Having looked at both versions, I would agree that the nested loop is easier to understand, so preferable, but I leave it up to you to request the change.

Reply:
Thanks for explanation, I prefer to request the change to remove "recursion"

Reply:
Frankly, I don't see how nested loops make code more readable. But ok.